### PR TITLE
Take unitTotal of order item to check if taxon rule can be applied

### LINKED ISSUE #9593: Promotion integrity error in total price of items from taxon
**Sylius version affected**: 1.1.6

**Description**  
Promotion integrity error when Sylius tries to choose between 2 promotions applyable in the same order

**Steps to reproduce**  

Create 2 Promotions:

First promotion: 
Rule1: At least 2 elements in cart
Rule2: At least 50€ in Taxon ABC

Action: 20% Discount in product of Taxon ABC
Promotion is exclusive.
Priority: 1

Second Promotion:
Rule1: At last 2 elements in cart
Rule2: At least 80€ in Taxon ABC

Action: 40% Discount in product of Taxon ABC
Promotion is exclusive.
Priority 2

If you insert 85€ in cart without discount, in review cart page you see that sylius apply promotion to cart and the total cart is 68€ (-20%).

When you want complete the payment, sylius now check the cart rules with the total amount and then launch an error with message:

promotion_integrity: 'You are no longer eligible for this promotion %promotionName%.' 

**Possible Solution**

### PR DESCRIPTION
**Pull request by jdeveloper**
| Q               | A
| --------------- | -----
| Branch?         | master 
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes #9593, #9512 
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php [hunk 1]
```diff
@@ -42,15 +49,18 @@
 
     /**
      * @param PromotionEligibilityCheckerInterface $promotionEligibilityChecker
+     * @param PromotionApplicatorInterface $promotionApplicator
      * @param EventDispatcherInterface $eventDispatcher
      * @param RouterInterface $router
      */
     public function __construct(
         PromotionEligibilityCheckerInterface $promotionEligibilityChecker,
+        PromotionApplicatorInterface $promotionApplicator,
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php]: 🚨 BC Break

File: features/checkout/order_promotion/order_promotion_integrity_validation.feature [hunk 2]
```diff
@@ -1,4 +1,4 @@
-@checkout
+@checkout @promotion_integrity
```
Comment (lchrusciel) [inline features/checkout/order_promotion/order_promotion_integrity_validation.feature]: Not needed as we didn't change anything in behat suites configuration

File: features/checkout/order_promotion/order_promotion_integrity_validation.feature [hunk 3]
```diff
@@ -16,9 +16,19 @@ Feature: Order promotions integrity
...
+        Given this promotion gives "$10.00" discount to every order
+        And this promotion expires tomorrow
+        And I added product "PHP T-Shirt" to the cart
         And I have proceeded selecting "Offline" payment method
         And this promotion has already expired
         When I confirm my order
         Then I should be informed that this promotion is no longer applied
         And I should not see the thank you page
+
+    @ui @promo_validation
```
Comment (lchrusciel) [inline features/checkout/order_promotion/order_promotion_integrity_validation.feature]: `@promo_validation` not needed as we didn't change anything in behat suites configuration

File: features/checkout/order_promotion/order_promotion_integrity_validation.feature [hunk 4]
```diff
@@ -16,9 +16,19 @@ Feature: Order promotions integrity
 
     @ui
     Scenario: Preventing customer from completing checkout with already expired promotion
-        Given I added product "PHP T-Shirt" to the cart
+        Given this promotion gives "$10.00" discount to every order
```
Comment (lchrusciel) [inline features/checkout/order_promotion/order_promotion_integrity_validation.feature]: Not needed as specified in background

File: features/checkout/order_promotion/order_promotion_integrity_validation.feature [hunk 5]
```diff
@@ -16,9 +16,19 @@ Feature: Order promotions integrity
...
+        And this promotion expires tomorrow
```
Comment (lchrusciel) [inline features/checkout/order_promotion/order_promotion_integrity_validation.feature]: Not needed as specified in background

File: src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php (hunk 1 above)
Comment (pamil) [reply src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php]: The solution here is to add it as the last argument, make it nullable, do `@trigger_error('deprecation message', \E_USER_DEPRECATED)` if that argument is null and do not run added logic if this promotion applicator is not there.

Comment (jdeveloper) [reply src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php]: @pamil Ok I'll make that change but do not now well what message to put. My english is not so well but if you give me a message suggestion I'll make the change.

**Comment by jdeveloper** [comment]
@lchrusciel @pamil thnaks for your suggestions. Pushed another commit. I hope it will be usefull.

File: src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php [hunk 6]
```diff
@@ -44,15 +51,22 @@
...
      * @param EventDispatcherInterface $eventDispatcher
      * @param RouterInterface $router
+     * @param PromotionApplicatorInterface $promotionApplicator
      */
     public function __construct(
         PromotionEligibilityCheckerInterface $promotionEligibilityChecker,
         EventDispatcherInterface $eventDispatcher,
-        RouterInterface $router
+        RouterInterface $router,
+        PromotionApplicatorInterface $promotionApplicator = null
```
Comment (pamil) [inline src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php]: `PromotionApplicatorInterface` -> `?PromotionApplicatorInterface`

File: src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php [hunk 7]
```diff
@@ -65,7 +79,18 @@ public function check(ResourceControllerEvent $event): void
 
         Assert::isInstanceOf($order, OrderInterface::class);
 
-        $promotions = $order->getPromotions();
+        // we creae a new promotion collection and remove them from cart
+        // so we can verify with original conditions (without the price beeng applied befor check)
```
Comment (pamil) [inline src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php]: Just a few typos here:
`creae` -> `create`
`beeng` -> `being`
`befor` -> `before`

File: src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php [hunk 8]
```diff
@@ -65,7 +79,18 @@ public function check(ResourceControllerEvent $event): void
...
+
+        $promotions = new ArrayCollection($order->getPromotions()->toArray());
```
Comment (pamil) [inline src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php]: I guess we don't need to put that array inside `ArrayCollection` again, `$order->getPromotions()->toArray()` would be enough.

**Comment by jdeveloper** [comment]
@pamil thanks, fixed and pushed again

File: src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php [hunk 9]
```diff
@@ -57,7 +56,7 @@ public function __construct(
         PromotionEligibilityCheckerInterface $promotionEligibilityChecker,
         EventDispatcherInterface $eventDispatcher,
         RouterInterface $router,
-        PromotionApplicatorInterface $promotionApplicator = null
+        ?PromotionApplicatorInterface $promotionApplicator
```
Comment (pamil) [inline src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php]: One more thing, that default `null` value should stay as is (so it's not breaking backwards compatibility)

Comment (jdeveloper) [reply src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php]: Done, sorry for missunderstanding you

**Comment by pamil** [comment]
Thank you, @jdeveloper! 🎉

**Comment by Chrysweel** [comment]
@pamil one question ! We have a project in production in version 1.1.6 and symfony 3.4.11

As this problem is a bug, will be update 1.1.x with this hotfix included ?

**Comment by pamil** [comment]
@Chrysweel as it introduces a deprecation in public API, we can't include it in a patch release for v1.1.x or v1.2.x branch. It will be included in v1.3.0, which is planned for the late September. If you need this fix ASAP, the best solution will be to overwrite the `OrderPromotionIntegrityChecker` with your own, which includes changes made in this PR.

**Comment by jdeveloper** [comment]
@pamil I don't understand well why it can't be included in v1.1.x or v1.2.x since it has an optional argument in the constructor and in case it is null, the previous behavior remains. You could use it in the same manner as before.

**Comment by pamil** [comment]
@jdeveloper https://semver.org/#spec-item-7
> Minor version Y (x.Y.z | x > 0) MUST be incremented if new, backwards compatible functionality is introduced to the public API. It MUST be incremented if any public API functionality is marked as deprecated.

**Comment by Chrysweel** [comment]
aha ok @pamil thanks for the aclaration.

shouldn't this issue be tagged with v1.3.0 to include in millestone?

https://github.com/Sylius/Sylius/milestone/19

**Pull request merged**